pyscf/cc/bccd.py

```diff
@@ -0,0 +1,263 @@
+        xcos = (rd[i, i] + rd[i + 1, i + 1]) * 0.5
+        xsin = (rd[i, i + 1] - rd[i + 1, i]) * 0.5
+        theta = numpy.arctan2(xsin, xcos)
+        ld[i, i + 1] = theta
+        ld[i + 1, i] = -theta
+    return rv @ ld @ rv.T
+
+def bccd_kernel_(mycc, u=None, conv_tol_normu=1e-5, max_cycle=20, diis=True,
+              verbose=4):
+    """
+    Brueckner coupled-cluster wrapper, using an outer-loop algorithm.
+    
+    Args:
+        mycc: a converged CCSD object.
+        u: initial transformation matrix.
+        conv_tol_normu: convergence tolerance for u matrix.
+        max_cycle: Maximum number of BCC cycles.
+        diis: whether perform DIIS.
+        verbose: verbose for CCSD inner iterations.
+
+    Returns:
+        mycc: a modified CC object with t1 vanished.
+    """ 
+    log = lib.logger.new_logger(mycc, verbose)
+    log.info("BCCD loop starts.")
+    
+    def trans_mo(mo_coeff, u):
+        mo_coeff = numpy.asarray(mo_coeff)
+        if mo_coeff.ndim == 2:
+            res = numpy.dot(mo_coeff, u)
+        else:
+            spin, nao, nmo = mo_coeff.shape
+            res = numpy.zeros((spin, nao, nmo), dtype=mo_coeff.dtype)
+            for s in range(spin):
+                res[s] = numpy.dot(mo_coeff[s], u[s])
+        return res
+    
+    def u2A(u):
+        if u.ndim == 2:
+            if la.det(u) < 0:
+                u[:, 0] *= -1
+            A = logm(u)
+        else:
+            if la.det(u[0]) < 0:
+                u[0][:, 0] *= -1
+            A_a = logm(u[0])
+            if la.det(u[1]) < 0:
+                u[1][:, 0] *= -1
+            A_b = logm(u[1])
+            A = numpy.asarray((A_a, A_b))
+        return A
+    
+    def A2u(A):
+        if A.ndim == 2:
+            u = la.expm(A)
+        else:
+            u_a = la.expm(A[0])
+            u_b = la.expm(A[1])
+            u = numpy.asarray((u_a, u_b))
+        return u
+    
+    mf = mycc._scf
+    ovlp = mf.get_ovlp()
+    adiis = lib.diis.DIIS()
+    frozen = mycc.frozen
+    level_shift = mycc.level_shift
+    
+    if u is None:
+        u = get_umat_from_t1(mycc.t1)
+    mo_coeff_ref = numpy.array(mf.mo_coeff, copy=True)
+    if u.ndim == 2:
+        u_tot = numpy.eye(u.shape[-1])
+    else:
+        u_tot = numpy.asarray((numpy.eye(u.shape[-1]), numpy.eye(u.shape[-1])))
+
+    with lib.temporary_env(mf, verbose=verbose):
+        e_tot_last = mycc.e_tot
+        for i in range(max_cycle):
+            u_tot = trans_mo(u_tot, u)
+            if diis:
+                A = u2A(u_tot)
+                if u_tot.ndim == 2:
+                    A_new = adiis.update(A, xerr=mycc.t1)
+                else:
+                    t1_ravel = numpy.hstack((mycc.t1[0].ravel(), mycc.t1[1].ravel()))
+                    A_new = adiis.update(A, xerr=t1_ravel)
+                u_tot = A2u(A_new)
+                mo_coeff_new = trans_mo(mo_coeff_ref, u_tot)
+                u = get_mo_ovlp(mf.mo_coeff, mo_coeff_new, ovlp)
```

Comment:
This line (208) does not appear to do anything, was it a debug test. *u* should tend to the unit matrix if things are going OK. The *else* statement is unnecessary as line 207 is common to if block and the else block (line 210).